winnowing.py

```diff
@@ -3,122 +3,123 @@
+                    newCode = newCode + code[mergedPoints[i][1] : mergedPoints[i+1][0]]
+                else:
+                    newCode = newCode + code[mergedPoints[i][1] :]
+        return {"ratio":(plagCount/len(code)),"Code":newCode}
 
-#we form windows of hash values and use min-hash to limit the number of fingerprints
-def fingerprints(arr, winSize = 4):
-    arrLen = len(arr)
-    prevMin = 0
-    currMin = 0
-    windows = []
-    fingerprintList = []
-    for i in range(arrLen - winSize):
-        win = arr[i: i + winSize]  #forming windows
-        windows.append(win)
-        currMin = i + minIndex(win)
-        if not currMin == prevMin:  #min value of window is stored only if it is not the same as min value of prev window
-            fingerprintList.append(arr[currMin])  #reduces the number of fingerprints while maintaining guarantee
-            prevMin = currMin  #refer to density of winnowing and guarantee threshold (Stanford paper)
+    #we form windows of hash values and use min-hash to limit the number of fingerprints
+    def fingerprints(self,arr, winSize = 4):
+        arrLen = len(arr)
+        prevMin = 0
+        currMin = 0
+        windows = []
+        fingerprintList = []
+        for i in range(arrLen - winSize):
+            win = arr[i: i + winSize]  #forming windows
+            windows.append(win)
+            currMin = i + self.minIndex(win)
+            if not currMin == prevMin:  #min value of window is stored only if it is not the same as min value of prev window
+                fingerprintList.append(arr[currMin])  #reduces the number of fingerprints while maintaining guarantee
+                prevMin = currMin  #refer to density of winnowing and guarantee threshold (Stanford paper)
 
-    return fingerprintList
+        return fingerprintList
+    #takes k-gram list as input and returns a list of only hash values
+    def hashList(self,arr):
+        HL = []
+        for i in arr:
+            HL.append(i[1])
 
-#takes k-gram list as input and returns a list of only hash values
-def hashList(arr):
-    HL = []
-    for i in arr:
-        HL.append(i[1])
+        return HL
+    #function to form k-grams out of the cleaned up text
+    def kgrams(self,text, k = 25):
+        tokenList = list(text)
+        n = len(tokenList)
+        kgrams = []
+        for i in range(n - k + 1):
+            kgram = ''.join(tokenList[i : i + k])
+            hv = self.hash(kgram)
+            kgrams.append((kgram, hv, i, i + k))  #k-gram, its hash value, starting and ending positions are stored
+            #these help in marking the plagiarized content in the original code.
+        return kgrams
+    #function that returns the index at which minimum value of a given list (window) is located
+    def minIndex(self,arr):
+        minI = 0
+        minV = arr[0]
+        n = len(arr)
+        for i in range(n):
+            if arr[i] < minV:
+                minV = arr[i]
+                minI = i
```

Comment:
No need of doing  `len(arr)` and `range(n)`. Get the list values directly.

Reply:
solved sir

Reply:
No need of for loop get the list minimum from min function and get the index of that value.